app/features/modelling/galaxy-builder/postProcessingRegl.js

```diff
@@ -0,0 +1,194 @@
+
+const asinhStretch = a => `
+float asinh(float px) {
+  return log(px + sqrt(1.0 + (px * px)));
+}
+float asinhStretch(float px) {
+  return asinh(px / ${a}) / asinh(${a});
+}
+`;
+
+export const convolvePSF = (r, psf, size) => {
+  const psfKeys = {};
+  let convolution = '';
+  let c = [0, 0];
+  for (let i = 0; i < size[0]; i += 1) {
+    for (let j = 0; j < size[1]; j += 1) {
```

Comment:
Maybe `forEach`?

Reply:
I'd like to keep this as is, I think it's more performant and I'm not sure nested forEach's would be much more readable?
```javascript
psf.forEach(
  (row, i) => row.forEach(
    (value, j) => {
      psfKeys[`psf[${(i * size[0]) + j}]`] = value;
      c = [i - 5, j - 5];
      convolution += `texture2D(texture, uv + vec2(${(c[0]).toFixed(1)}/texSize[0],`;
      convolution += `${(c[1]).toFixed(1)}/texSize[1])) * psf[${(i * size[0]) + j}]`;
      if (i < size[0] - 1 || j < size[1] - 1) {
        convolution += ' + \n';
      }
    }
  )
)
```

Reply:
For loops are faster, though, typically the difference (in milliseconds) won't be noticed by the end user and performance is still variable with the browser. If it can be demonstrated that there is a noticeable impact, using say dev tool's performance analysis tools, then we can make a decisions based on performance. 

Generally the newer array enumerables are preferred since that's the direction javascript is moving and I do think readability, although subjective, is an important consideration for the team. I find seeing named variables like `row` and `value` more immediately readable at a glance than the for loop.

Reply:
Good point, premature optimisation is the root of all evil and all that. It also only runs once in a classification rather than once a frame, which I'd forgotten! I feel like this should be commented too as it's about as readable as Daily Mail written backwards in webdings... I'll add this to the list of things to put in my follow-up PR

Reply:
Agreed re. readability being paramount. The for loop obsession comes from when I was doing rendering on the CPU about a year ago and was desperately trying to scrape back every possible millisecond.